packages/vite/src/index.ts

```diff
@@ -0,0 +1,110 @@
+  rollupOptions?: AllowedRollupOptions;
+};
+
+type AllowedViteConfig = Omit<InlineConfig, 'base' | 'build' | 'mode' | 'resolve' | 'root'> & {
+  build?: AllowedBuildOptions;
+};
+
+interface Options {
+  viteConfig: AllowedViteConfig;
+}
+
+export class VitePackager implements PackagerInstance {
+  static annotation = '@embroider/vite';
+
+  private viteConfig: AllowedViteConfig;
+  private variant: Variant;
+
+  pathToVanillaApp: string;
+
+  constructor(
+    pathToVanillaApp: string,
+    private outputPath: string,
+    variants: Variant[],
+    _consoleWrite: (msg: string) => void,
+    options?: Options
+  ) {
+    this.pathToVanillaApp = realpathSync(pathToVanillaApp);
+
+    // For now we're not worried about building each variant
+    // Let's just assume we have one
+    this.variant = variants[0];
+
+    this.viteConfig = options?.viteConfig ?? {};
+  }
+
+  async build(): Promise<void> {
+    const meta = this.getAppMeta();
+
+    const entryPoints = meta.assets
+      .map(asset => parse(asset))
+      .filter(asset => asset.ext === '.html')
+      .map(asset => join(this.pathToVanillaApp, asset.dir, asset.base));
+
+    await build({
+      ...this.viteConfig,
+      mode: this.variant.optimizeForProduction ? 'production' : 'development',
+      logLevel: 'error',
+      plugins: [
+        templateCompilerPlugin({
+          templateCompilerFile: join(this.pathToVanillaApp, meta['template-compiler'].filename),
+          variant: this.variant,
+        }),
+        babel({
+          // Embroider includes the Runtime plugin in the generated Babel config
+          babelHelpers: 'runtime',
+
+          // Path to the Embroider-generated Babel config
+          configFile: join(this.pathToVanillaApp, meta.babel.filename),
+
+          // Path to the Embroider-generated file defining a filtering function
+          // eslint-disable-next-line @typescript-eslint/no-require-imports
+          filter: require(join(this.pathToVanillaApp, meta.babel.fileFilter)),
+        }),
+        commonjs(),
+        ...(this.viteConfig.plugins ?? []),
+      ],
+      root: this.pathToVanillaApp,
+      base: meta['root-url'],
+      resolve: {
+        extensions: meta['resolvable-extensions'],
+      },
+      build: {
+        ...this.viteConfig.build,
+        outDir: this.outputPath,
+        rollupOptions: {
+          ...this.viteConfig.build?.rollupOptions,
+          input: entryPoints,
+        },
+      },
+    });
+
+    // Copy over the `vendor.js` file that is ignored by Vite
+    await this.passThrough('assets/vendor.js');
+    await this.passThrough('assets/vendor.map');
+  }
+
+  private getAppMeta(): AppMeta {
+    return JSON.parse(readFileSync(join(this.pathToVanillaApp, 'package.json'), 'utf8'))['ember-addon'] as AppMeta;
+  }
+
+  private async passThrough(path: string) {
+    const source = join(this.pathToVanillaApp, path);
+    const dest = join(this.outputPath, path);
+
+    await copyFile(source, dest);
+  }
```

Comment:
These are both functionalities that are more-or-less copied from the Webpack packager. Maybe they belong on a common abstract base class that can be extended from, rather than an interface that is implemented?

The whole HTML entry point logic from the Webpack plugin might be helpful here, honestly; especially the discovery of non-module scripts that need to be manually handled by both of the Embroider packages.

Reply:
Yes, we can extract and share the html utilities.

Reply:
Cool! I did some refactoring last night to facilitate this.

`Packager` is now an abstract base class instead of an interface, which gives a place for us to put some shared functionality for packagers. Some of the parts of the Webpack package have now moved to Core, but it could be worth making some kind of separate Packager packager for that stuff instead. 

If it's helpful, I can totally create a separate PR for refactoring that code to keep it apart from adding a new packager.

Reply:
> Packager is now an abstract base class instead of an interface,

I would rather keep it an interface and put shared code in a utility package. Because I don't think all packagers will want this stuff. I would hope that some will, for example, be better at handling standards-compliant HTML entrypoints directly, without us needing to do so much to read the HTML for them.